### PR TITLE
Cold notifiers -- if a cat mews but nobody hears it, does it make a sound?

### PR DESCRIPTION
**Pull request by michaelfig**
# Context

Notifiers are great because they don't make sounds unless somebody is waiting to hear them.  However, "hot" notifiers think furiously to themselves until somebody explicitly stops them (if ever).  "Cold" notifiers only think immediately before they make a sound, and peacefully cease to exist when they are forgotten.

In the Observable literature, this is the difference between a "hot" observable (which is always active, regardless of whether it has observers), and a "cold" observable which is only active when it is being observed.

Our existing notifier mechanism is useful, but only supports "hot" notifiers.  We can solve this without needing to lean on big hammers like WeakRefs.

# The Cold Notifier implementation

This PR introduces a capability as part of the NotifierKit that allows updater code to register an "onObserved" callback.  By doing so, the updater can be quiescent until somebody calls `E(notifier).getUpdateSince(nextStateCount)`, when `onObserved(nextStateCount)` will be called.

A "cold" notifier can be built by only subscribing to resource events when a call to `onObserved` callback implies that somebody is watching.  This way, garbage collection is still not observable by the updater, but it can, when nobody is listening, be garbage-collected without any resource leaks.

The difference to the caller is that the `updateCount` is only incremented once in a while, instead of every `interval`.  As soon as the notifier is not asked for updates, the timer updater quiesces.  If a timer notifier has no external references, once the updater has also quiesced, the entire reference graph for the notifier and updater and `onObserved` callback can be garbage-collected without leaving behind any eternal subscriptions.

# Use cases

As a motivating example in this PR, the TimerService `E(timerService).makeNotifier(delay, interval)` currently returns a "hot" notifier that holds onto an eternal subscription to the timer device, causing both inefficiency and a resource leak.  This PR reimplements that API with a "cold" notifier which only subscribes to individual timer wake-ups when their presence is required by actual observations.

In a future PR to solve #3629, we'd like to use a cold notifier for the "bank purse" balance notifiers.  That way, the Golang side of the bank purses won't actually need to publish balance changes unless there is somebody is waiting on the corresponding JS balance notifiers.  This will usefully reduce the amount of messages to SwingSet when transfers are made on the Cosmos side.

### CODE REVIEW

**Comment by erights** [comment]
@michaelfig this is still Draft. Should I wait before starting to review?

**Review by erights — CHANGES_REQUESTED** [review]
Could use a test ;)

File: packages/notifier/src/notifier.js [hunk 1]
```diff
@@ -77,7 +82,36 @@ export const makeNotifierKit = (...args) => {
...
+        // Queue the callbacks in a later turn so that they can't interfere with
+        // our plans.
+        const observedUpdateCount = lastObservedUpdateCount;
+        Promise.resolve().then(() => {
+          if (!onObservedCallbacks) {
+            return;
+          }
+          onObservedCallbacks.forEach(cb => {
+            try {
+              cb(observedUpdateCount);
```
Comment (erights) [inline packages/notifier/src/notifier.js:103]: You're calling these synchronously as you're iterating the set. What happens if one calls `registerOnObserved` during the iteration? What happens if a cancel is invoked during the iteration?

**Review by erights — CHANGES_REQUESTED** [review]
I propose https://github.com/Agoric/agoric-sdk/pull/3949 as an alternative to this PR. @michaelfig, as my reviewer on that one, if I missed something from this PR, please let me know.

**Comment by michaelfig** [comment]
> I propose #3949 as an alternative to this PR. @michaelfig, as my reviewer on that one, if I missed something from this PR, please let me know.

I like your alternative much better.  Closing this PR in favour of that one.

**Pull request closed without merging**